models/mask_rcnn.py

```diff
@@ -0,0 +1,247 @@
+
+
+def generate_anchors(
+    stride=16, sizes=(32, 64, 128, 256, 512), aspect_ratios=(0.5, 1, 2)
+):
+  return _generate_anchors(stride, np.array(sizes, dtype=np.float32) / stride, np.array(aspect_ratios, dtype=np.float32))
+
+
+def _generate_anchors(base_size, scales, aspect_ratios):
+  anchor = np.array([1, 1, base_size, base_size], dtype=np.float32) - 1
+  anchors = _ratio_enum(anchor, aspect_ratios)
+  anchors = np.vstack(
+    [_scale_enum(anchors[i, :], scales) for i in range(anchors.shape[0])]
+  )
+  return anchors
+
+
+def _whctrs(anchor):
+  w = anchor[2] - anchor[0] + 1
+  h = anchor[3] - anchor[1] + 1
+  x_ctr = anchor[0] + 0.5 * (w - 1)
+  y_ctr = anchor[1] + 0.5 * (h - 1)
+  return w, h, x_ctr, y_ctr
+
+
+def _mkanchors(ws, hs, x_ctr, y_ctr):
+  ws = ws[:, np.newaxis]
+  hs = hs[:, np.newaxis]
+  anchors = np.hstack((
+    x_ctr - 0.5 * (ws - 1),
+    y_ctr - 0.5 * (hs - 1),
+    x_ctr + 0.5 * (ws - 1),
+    y_ctr + 0.5 * (hs - 1),
+  ))
+  return anchors
+
+
+def _ratio_enum(anchor, ratios):
+  w, h, x_ctr, y_ctr = _whctrs(anchor)
+  size = w * h
+  size_ratios = size / ratios
+  ws = np.round(np.sqrt(size_ratios))
+  hs = np.round(ws * ratios)
+  anchors = _mkanchors(ws, hs, x_ctr, y_ctr)
+  return anchors
+
+
+def _scale_enum(anchor, scales):
+  w, h, x_ctr, y_ctr = _whctrs(anchor)
+  ws = w * scales
+  hs = h * scales
+  anchors = _mkanchors(ws, hs, x_ctr, y_ctr)
+  return anchors
+
+
+class RPNHead:
+  def __init__(self, in_channels, num_anchors):
+    self.conv = nn.Conv2d(in_channels, 256, kernel_size=3, padding=1)
+    self.cls_logits = nn.Conv2d(256, num_anchors, kernel_size=1)
+    self.bbox_pred = nn.Conv2d(256, num_anchors * 4, kernel_size=1)
+
+
+class RPN:
+  def __init__(self, in_channels):
+    self.anchor_generator = AnchorGenerator()
+    self.head = RPNHead(in_channels, self.anchor_generator.num_anchors_per_location()[0])
+
+
+def make_conv3x3(
+  in_channels,
+  out_channels,
+  dilation=1,
+  stride=1,
+  use_gn=False,
+  use_relu=False,
+  kaiming_init=True
+):
+  conv = nn.Conv2d(
+    in_channels,
+    out_channels,
+    kernel_size=3,
+    stride=stride,
+    padding=dilation,
+    dilation=dilation,
+    bias=False if use_gn else True
+  )
+  return conv
+
+
+class MaskRCNNFPNFeatureExtractor:
+  def __init__(self):
+    resolution = 14
+    scales = (1.0 / 16,)
+    sampling_ratio = 0
+    pooler = Pooler(
+      output_size=(resolution, resolution),
+      scales=scales,
+      sampling_ratio=sampling_ratio,
+    )
+    input_size = 256
+    self.pooler = pooler
+
+    use_gn = False
+    layers = (256, 256, 256, 256)
+    dilation = 1
+
+    next_feature = input_size
+    self.blocks = []
+    for layer_idx, layer_features in enumerate(layers, 1):
+      layer_name = "mask_fcn{}".format(layer_idx)
+      module = make_conv3x3(next_feature, layer_features,
+                            dilation=dilation, stride=1, use_gn=use_gn
+                            )
+      exec(f"self.{layer_name} = module")
```

Comment:
this is kinda cursed, you should be able to change the name during the weight loading process.

Reply:
@wozeparrot yes, a lot of things are cursed right now, will be taking this up one by one, while I'm adding the calls